File: auto_outsrc/parser/HVE/HVE_LATEST.cpp
#include "Charm.h"
#include <iostream>
#include <sstream>
#include <string>
#include <list>
using namespace std;

PairingGroup group(AES_SECURITY);

void setup(int n, CharmList & pk, CharmList & msk)
{
    G1 g1 = group.init(G1_t);
    G2 g2 = group.init(G2_t);
    GT egg = group.init(GT_t);
    ZR y = group.init(ZR_t);
    GT Y = group.init(GT_t);
    CharmListZR t;
    CharmListZR v;
    CharmListZR r;
    CharmListZR m;
    CharmListG1 T;
    CharmListG1 V;
    CharmListG1 R;
    CharmListG1 M;
    g1 = group.random(G1_t);
    g2 = group.random(G2_t);
    egg = group.pair(g1, g2);
    y = group.random(ZR_t);
    Y = group.exp(egg, y);
    for (int i = 0; i < n; i++)
    {
        t.insert(i, group.random(ZR_t));
        v.insert(i, group.random(ZR_t));
        r.insert(i, group.random(ZR_t));
        m.insert(i, group.random(ZR_t));
        T.insert(i, group.exp(g1, t[i]));
        V.insert(i, group.exp(g1, v[i]));
        R.insert(i, group.exp(g1, r[i]));
        M.insert(i, group.exp(g1, m[i]));
    }
    pk.insert(0, g1);
    pk.insert(1, g2);
    pk.insert(2, Y);
    pk.insert(3, T);
    pk.insert(4, V);
    pk.insert(5, R);
    pk.insert(6, M);
    msk.insert(0, y);
    msk.insert(1, t);
    msk.insert(2, v);
    msk.insert(3, r);
    msk.insert(4, m);
    return;
}

void keygen(CharmList & pk, CharmList & msk, CharmListInt & yVector, ZR & bf0, CharmList & skBlinded)
{
    G1 g1;
    G2 g2;
    GT Y;
    CharmListG1 T;
    CharmListG1 V;
    CharmListG1 R;
    CharmListG1 M;
    ZR y;
    CharmListZR t;
    CharmListZR v;
    CharmListZR r;
    CharmListZR m;
    int numNonDontCares = 0;
    ZR sumUSaisUSsoFar = group.init(ZR_t, 0);
    int endForLoop = 0;
    CharmListZR a;
    int currentUSaUSindex = 0;
    CharmListG2 YVector;
    CharmListG2 LVector;
    CharmListZR blindingFactorYVectorBlinded;
    CharmListG2 YVectorBlinded;
    CharmListZR blindingFactorLVectorBlinded;
    CharmListG2 LVectorBlinded;
    bf0 = group.random(ZR_t);
    
    g1 = pk[0].getG1();
    g2 = pk[1].getG2();
    Y = pk[2].getGT();
    T = pk[3].getListG1();
    V = pk[4].getListG1();
    R = pk[5].getListG1();
    M = pk[6].getListG1();
    
    y = msk[0].getZR();
    t = msk[1].getListZR();
    v = msk[2].getListZR();
    r = msk[3].getListZR();
    m = msk[4].getListZR();
    //;
    for (int i = 0; i < n; i++)
    {
        if ( ( (yVector[i]) != (2) ) )
        {
            numNonDontCares = group.add(numNonDontCares, 1);
        }
    }
    group.init(sumUSaisUSsoFar, 0);
    endForLoop = group.sub(numNonDontCares, 1);
    for (int i = 0; i < endForLoop; i++)
    {
        a.insert(i, group.random(ZR_t));
        sumUSaisUSsoFar = group.add(sumUSaisUSsoFar, a[i]);
    }
    a.insert(numNonDontCares-1, group.sub(y, sumUSaisUSsoFar));
    //;
    for (int i = 0; i < n; i++)
    {
        if ( ( (yVector[i]) == (2) ) )
        {
            YVector.insert(i, group.init(G2_t));
            LVector.insert(i, group.init(G2_t));
        }
        if ( ( (yVector[i]) == (0) ) )
        {
            YVector.insert(i, group.exp(g2, group.div(a[currentUSaUSindex], r[i])));
            LVector.insert(i, group.exp(g2, group.div(a[currentUSaUSindex], m[i])));
            currentUSaUSindex = group.add(currentUSaUSindex, 1);
        }
        if ( ( (yVector[i]) == (1) ) )
        {
            YVector.insert(i, group.exp(g2, group.div(a[currentUSaUSindex], t[i])));
            LVector.insert(i, group.exp(g2, group.div(a[currentUSaUSindex], v[i])));
            currentUSaUSindex = group.add(currentUSaUSindex, 1);
        }
    }
    CharmList YVector_keys = YVector.keys();
    int YVector_len = YVector.length();
    for (int y_var = 0; y_var < YVector_len; y_var++)
    {
        y = YVector_keys[y_var];
        blindingFactorYVectorBlinded.insert(y, bf0);
        YVectorBlinded.insert(y, group.exp(YVector[y], group.div(1, blindingFactorYVectorBlinded[y].getZR())));
    }
    CharmList LVector_keys = LVector.keys();
    int LVector_len = LVector.length();
    for (int y_var = 0; y_var < LVector_len; y_var++)
    {
        y = LVector_keys[y_var];
        blindingFactorLVectorBlinded.insert(y, bf0);
        LVectorBlinded.insert(y, group.exp(LVector[y], group.div(1, blindingFactorLVectorBlinded[y].getZR())));
    }
    skBlinded.insert(0, YVectorBlinded);
    skBlinded.insert(1, LVectorBlinded);
    return;
}

void encrypt(GT & Message, CharmListInt & xVector, CharmList & pk, CharmList & CT)
{
    G1 g1;
    G2 g2;
    GT Y;
    CharmListG1 T;
    CharmListG1 V;
    CharmListG1 R;
    CharmListG1 M;
    ZR s = group.init(ZR_t);
    CharmListZR sUSi;
    GT omega = group.init(GT_t);
    G1 C0 = group.init(G1_t);
    CharmListG1 XVector;
    CharmListG1 WVector;
    
    g1 = pk[0].getG1();
    g2 = pk[1].getG2();
    Y = pk[2].getGT();
    T = pk[3].getListG1();
    V = pk[4].getListG1();
    R = pk[5].getListG1();
    M = pk[6].getListG1();
    s = group.random(ZR_t);
    for (int i = 0; i < n; i++)
    {
        sUSi.insert(i, group.random(ZR_t));
    }
    omega = group.mul(Message, group.exp(Y, group.neg(s)));
    C0 = group.exp(g1, s);
    for (int i = 0; i < n; i++)
    {
        if ( ( (xVector[i]) == (0) ) )
        {
            XVector.insert(i, group.exp(R[i], group.sub(s, sUSi[i])));
            WVector.insert(i, group.exp(M[i], sUSi[i]));
        }
        if ( ( (xVector[i]) == (1) ) )
        {
            XVector.insert(i, group.exp(T[i], group.sub(s, sUSi[i])));
            WVector.insert(i, group.exp(V[i], sUSi[i]));
        }
    }
    CT.insert(0, omega);
    CT.insert(1, C0);
    CT.insert(2, XVector);
    CT.insert(3, WVector);
    return;
}

void transform(CharmList & CT, CharmList & skBlinded, CharmList & transformOutputList, int nn)
{
    GT omega;
    G1 C0;
    CharmListG1 XVector;
    CharmListG1 WVector;
    CharmListG2 YVectorBlinded;
    CharmListG2 LVectorBlinded;
    G2 g2Id = group.init(G2_t);
    GT intermediateResults = group.init(GT_t);
    
    omega = CT[0].getGT();
    C0 = CT[1].getG1();
    XVector = CT[2].getListG1();
    WVector = CT[3].getListG1();
    
    YVectorBlinded = skBlinded[0].getListG2();
    LVectorBlinded = skBlinded[1].getListG2();
    transformOutputList.insert(1, omega);
    transformOutputList.insert(0, group.init(G2_t));
    g2Id = transformOutputList[0].getG2();
    nn = YVectorBlinded.length();
    for (int i = 0; i < nn; i++)
    {
		//NOP;
        if ( ( (( (YVectorBlinded[i]) != (g2Id) )) && (( (LVectorBlinded[i]) != (g2Id) )) ) )
        {
		//NOP;
            transformOutputList.insert(1000+7*i, group.mul(group.pair(XVector[i], YVectorBlinded[i]), group.pair(WVector[i], LVectorBlinded[i])));
            intermediateResults = transformOutputList[1000+7*i].getGT();
        }
    }
    return;
}

void decout(CharmList & transformOutputList, ZR & bf0, int nn, GT & Message2)
{
    GT omega = group.init(GT_t);
    GT dotProd = group.init(GT_t, 1);
    G2 g2Id = group.init(G2_t);
    GT intermediateResults = group.init(GT_t);
    omega = transformOutputList[1].getGT();
    group.init(dotProd, 1);
    g2Id = transformOutputList[0].getG2();
    for (int i = 0; i < nn; i++)
    {
		//NOP;
        if ( ( (( (YVectorBlinded[i]) != (g2Id) )) && (( (LVectorBlinded[i]) != (g2Id) )) ) )
        {
		//NOP;
            intermediateResults = group.exp(transformOutputList[1000+7*i].getGT(), bf0);
            dotProd = group.mul(dotProd, intermediateResults);
        }
    }
    Message2 = group.mul(omega, dotProd);
    return;
}

int main()
{
    return 0;
}

File: auto_outsrc/parser/HVE/HVE_LATEST.py
from HVE_LATEST_USER import *

from charm.toolbox.pairinggroup import PairingGroup,ZR,G1,G2,GT,pair
from charm.core.engine.util import *
from charm.core.math.integer import randomBits

group = None

secparam = 80


def setup(n):
    M = {}
    V = {}
    m = {}
    r = {}
    t = {}
    v = {}
    R = {}
    T = {}

    g1 = group.random(G1)
    g2 = group.random(G2)
    egg = pair(g1, g2)
    y = group.random(ZR)
    Y = (egg ** y)
    for i in range(0, n):
        t[i] = group.random(ZR)
        v[i] = group.random(ZR)
        r[i] = group.random(ZR)
        m[i] = group.random(ZR)
        T[i] = (g1 ** t[i])
        V[i] = (g1 ** v[i])
        R[i] = (g1 ** r[i])
        M[i] = (g1 ** m[i])
    pk = [g1, g2, Y, T, V, R, M, n]
    msk = [y, t, v, r, m]
    output = (pk, msk)
    return output

def keygen(pk, msk, yVector):
    a = {}
    LVectorBlinded = {}
    blindingFactorLVectorBlinded = {}
    blindingFactorYVectorBlinded = {}
    LVector = {}
    YVectorBlinded = {}
    YVector = {}

    bf0 = group.random(ZR)
    g1, g2, Y, T, V, R, M, n = pk
    y, t, v, r, m = msk
    numNonDontCares = 0
    for i in range(0, n):
        if ( ( (yVector[i]) != (2) ) ):
            numNonDontCares = (numNonDontCares + 1)
    sumUSaisUSsoFar = 0
    endForLoop = (numNonDontCares - 1)
    for i in range(0, endForLoop):
        a[i] = group.random(ZR)
        sumUSaisUSsoFar = (sumUSaisUSsoFar + a[i])
    a[numNonDontCares-1] = (y - sumUSaisUSsoFar)
    currentUSaUSindex = 0
    for i in range(0, n):
        if ( ( (yVector[i]) == (2) ) ):
            YVector[i] = group.init(G2)
            LVector[i] = group.init(G2)
        if ( ( (yVector[i]) == (0) ) ):
            YVector[i] = (g2 ** (a[currentUSaUSindex] / r[i]))
            LVector[i] = (g2 ** (a[currentUSaUSindex] / m[i]))
            currentUSaUSindex = (currentUSaUSindex + 1)
        if ( ( (yVector[i]) == (1) ) ):
            YVector[i] = (g2 ** (a[currentUSaUSindex] / t[i]))
            LVector[i] = (g2 ** (a[currentUSaUSindex] / v[i]))
            currentUSaUSindex = (currentUSaUSindex + 1)
    for y in YVector:
        blindingFactorYVectorBlinded[y] = bf0
        YVectorBlinded[y] = (YVector[y] ** (1 / blindingFactorYVectorBlinded[y]))
    for y in LVector:
        blindingFactorLVectorBlinded[y] = bf0
        LVectorBlinded[y] = (LVector[y] ** (1 / blindingFactorLVectorBlinded[y]))
    skBlinded = [YVectorBlinded, LVectorBlinded]
    output = (bf0, skBlinded)
    return output

def encrypt(Message, xVector, pk):
    WVector = {}
    sUSi = {}
    XVector = {}

    g1, g2, Y, T, V, R, M, n = pk
    s = group.random(ZR)
    for i in range(0, n):
        sUSi[i] = group.random(ZR)
    omega = (Message * (Y ** -s))
    C0 = (g1 ** s)
    for i in range(0, n):
        if ( ( (xVector[i]) == (0) ) ):
            XVector[i] = (R[i] ** (s - sUSi[i]))
            WVector[i] = (M[i] ** sUSi[i])
        if ( ( (xVector[i]) == (1) ) ):
            XVector[i] = (T[i] ** (s - sUSi[i]))
            WVector[i] = (V[i] ** sUSi[i])
    CT = [omega, C0, XVector, WVector]
    output = CT
    return output

def transform(CT, skBlinded):
    transformOutputList = {}

    omega, C0, XVector, WVector = CT
    YVectorBlinded, LVectorBlinded = skBlinded
    transformOutputList[1] = omega
    transformOutputList[0] = group.init(G2)
    g2Id = transformOutputList[0]
    nn = len(YVectorBlinded)
    for i in range(0, nn):
        pass
        if ( ( (( (YVectorBlinded[i]) != (g2Id) )) and (( (LVectorBlinded[i]) != (g2Id) )) ) ):
            pass
            transformOutputList[1000+7*i] = (pair(XVector[i], YVectorBlinded[i]) * pair(WVector[i], LVectorBlinded[i]))
            intermediateResults = transformOutputList[1000+7*i]
    output = (transformOutputList, nn)
    return output

def decout(transformOutputList, bf0, nn):
    omega = transformOutputList[1]
    dotProd = group.init(GT)
    g2Id = transformOutputList[0]
    for i in range(0, nn):
        pass
        if ( ( (( (YVectorBlinded[i]) != (g2Id) )) and (( (LVectorBlinded[i]) != (g2Id) )) ) ):
            pass
            intermediateResults = (transformOutputList[1000+7*i] ** bf0)
            dotProd = (dotProd * intermediateResults)
    Message2 = (omega * dotProd)
    output = Message2
    return output

def SmallExp(bits=80):
    return group.init(ZR, randomBits(bits))

def main():
    global group
    group = PairingGroup("SS512")

    (pk, msk) = setup(4)
    (bf0, skBlinded) = keygen(pk, msk, [2, 1, 0, 2])
    M = group.random(GT)
    print(M)
    print("\n\n")
    CT = encrypt(M, [1, 1, 0, 1], pk)
    (transformOutputList, nn) = transform(CT, skBlinded)
    M2 = decout(transformOutputList, bf0, nn)

    print(M2)
    if (M == M2):
        print("success")
    else:
        print("failed")


if __name__ == '__main__':
    main()

